### PR TITLE
Add postgres SQL support and LDAP configuration

### PR DESCRIPTION
**Pull request by clementTal**
# Add postgres SQL support by adding a PSQL database.
To run PSQL migration: 
`python manage.py migrate --database=psql`
Othervise default database (MySql) will be used

# LDAP configuration
Add LDAP config to environments vars and file. Doing this, we can add the ldap_config.py file on the docker image and repo.

# Use latest version of poetry
Update the poetry version used by the project

# Notes:
- changes can be tested using public docker image: https://hub.docker.com/r/clemtal/squest

### CODE REVIEW

**Comment by Sispheor** [comment]
Hi, thanks for the PR.

I'm not sure what to think about it. Usually I don't change piece of software used by project owner 😅. Like, you have certainly AWX or Tower behind and you've took the proposed database as it, right?
What is the goal of adding this support? What is the value added?

Our target will be to place Squest installation in K8S/OCP with an operator or something to deploy it. So the DB is a part of the deployment and should not be replaced.

File: .dockerignore [hunk 1]
```diff
@@ -1,7 +1,7 @@
 /static
 /node_modules
 /media
-*/ldap_config.py*
```
Comment (Sispheor) [inline .dockerignore:4]: The ldap config should be added by a volume and so not build in the image. As it may contains secrets...

File: .gitignore [hunk 2]
```diff
@@ -3,7 +3,7 @@ media/service_image
 media/portfolio_image
 media/doc_images
 static/
-ldap_config.py
```
Comment (Sispheor) [inline .gitignore:6]: Same as previous comment

File: .gitignore [hunk 3]
```diff
@@ -115,3 +115,5 @@ venv.bak/
 
 # ansible
 *.retry
+
+custom.*
```
Comment (Sispheor) [inline .gitignore:119]: what is that?

File: Squest/ldap_config.py [hunk 4]
```diff
@@ -6,18 +6,20 @@
 # -----------------------
```
Comment (Sispheor) [inline Squest/ldap_config.py:6]: This file is an example. Each user ned to configure it following his env. 
We don't use "env" variable as it would be hard to cover all flag proposed by the plugin.

File: docker/nginx.conf [hunk 5]
```diff
@@ -45,7 +45,7 @@ http {
     # squest app
     server django:8000; # fail_timeout=1s max_fails=1;
     # maintenance page
-    server maintenance:80 backup;
+    # server maintenance:80 backup;
```
Comment (Sispheor) [inline docker/nginx.conf:48]: Why removing this?

File: docker/scripts/mariadb-init.sh [hunk 6]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
```
Comment (Sispheor) [inline docker/scripts/mariadb-init.sh:1]: Why switching to a bash script?
The sql file is the way proposed in the mysql container doc.

File: docker-compose.yml [hunk 7]
```diff
@@ -57,6 +47,8 @@ services:
     image: nginx:alpine
     depends_on:
       - django
+    ports:
```
Comment (Sispheor) [inline docker-compose.yml:50]: We have a reverse proxy in front so we don't expose the internal 8000 port.

File: docker-compose.yml [hunk 8]
```diff
@@ -2,13 +2,6 @@
 version: '3.7'
 
 services:
-  db:
```
Comment (Sispheor) [inline docker-compose.yml:5]: Why not leaving it as default? And overriding only when postpres is used?

**Comment by clementTal** [comment]
> Hi, thanks for the PR.
> 
> I'm not sure what to think about it. Usually I don't change piece of software used by project owner 😅. 

It's the purpose of opensource, isn't it ?

> Like, you have certainly AWX or Tower behind and you've took the proposed database as it, right? What is the goal of adding this support? What is the value added?

Because AWX use already Postgres, and my company do not accept other databases.

> 
> Our target will be to place Squest installation in K8S/OCP with an operator or something to deploy it. So the DB is a part of the deployment and should not be replaced.

I do not share this point of view, as k8s and ocp are designed to run stateless services. Most projects uses k8s/OCP to host front and backend app, and use managed databases to host datas.

File: docker/scripts/mariadb-init.sh (hunk 6 above)
Comment (clementTal) [reply docker/scripts/mariadb-init.sh:1]: Using a bash script is an option too (see https://hub.docker.com/_/mariadb).
Doing this, we can change database user based on environement vars. Using your sql, we need to set user to "squest_user"

File: docker-compose.yml (hunk 7 above)
Comment (clementTal) [reply docker-compose.yml:50]: Does api calls works without this port open ?

File: .dockerignore (hunk 1 above)
Comment (clementTal) [reply .dockerignore:4]: It's part of the app, not config files. It should nevers contains secrets, as there should be externalized.

File: .gitignore (hunk 3 above)
Comment (clementTal) [reply .gitignore:119]: a way to pass custom docker files and envs files (should be ignored as it may contains secrets)

File: docker/nginx.conf (hunk 5 above)
Comment (clementTal) [reply docker/nginx.conf]: fixed

File: docker-compose.yml (hunk 8 above)
Comment (clementTal) [reply docker-compose.yml:5]: fixed

File: docker-compose.yml (hunk 7 above)
Comment (clementTal) [reply docker-compose.yml]: fixed

File: .gitignore (hunk 2 above)
Comment (clementTal) [reply .gitignore:6]: It's part of the app, not config files. It should nevers contains secrets, as there should be externalized.

File: Squest/ldap_config.py (hunk 4 above)
Comment (clementTal) [reply Squest/ldap_config.py:6]: It could be overided by users using volumes, but it should be included in repo with most of options covered by envs.

File: docs/dev/dev-env.md [hunk 9]
```diff
@@ -38,7 +38,11 @@ The development environment is composed of 4 parts:
 
 Run the Docker compose file with only required services to bring up database, message broker and other required system
 ```bash
-docker-compose -f docker-compose.yml -f dev.docker-compose.yml up db phpmyadmin rabbitmq redis-cache
+docker-compose -f docker-compose.yml  -f mysql.docker-composer.yaml -f dev.docker-compose.yml up db phpmyadmin rabbitmq redis-cache
```
Comment (Sispheor) [inline docs/dev/dev-env.md:41]: this approach will break the default behavior where we have only one compose. In the doc, new user can start Squest from a simple "docker-compose up".
So it would be better to add a side compose for postgres.

File: mysql.docker-composer.yaml [hunk 10]
```diff
@@ -0,0 +1,23 @@
+services:
```
Comment (Sispheor) [inline mysql.docker-composer.yaml:1]: This part should still be the default.

File: Squest/ldap_config.py (hunk 4 above)
Comment (Sispheor) [reply Squest/ldap_config.py:6]: So variables should be in the doc.
And both usage need to be documented.

File: .dockerignore (hunk 1 above)
Comment (Sispheor) [reply .dockerignore:4]: Need to be documented then

File: Squest/ldap_config.py [hunk 11]
```diff
@@ -6,18 +6,20 @@
...
+LDAP_CA_FILE_PATH = os.environ.get('LDAP_CA_FILE_PATH', "/usr/local/share/ca-certificates/ldap_ca.crt")
+
 AUTH_LDAP_CONNECTION_OPTIONS = {
     ldap.OPT_X_TLS_CACERTFILE: LDAP_CA_FILE_PATH,
-    ldap.OPT_X_TLS_REQUIRE_CERT: ldap.OPT_X_TLS_NEVER,
+    ldap.OPT_X_TLS_REQUIRE_CERT: ldap.OPT_X_TLS_ALLOW,
     ldap.OPT_X_TLS_NEWCTX: 0
 }
+
 AUTH_LDAP_USER_ATTR_MAP = {
```
Comment (Sispheor) [inline Squest/ldap_config.py:21]: Fields here should be wrapped as well then to be consistent. Like in [Netbox project](https://github.com/netbox-community/netbox-docker/blob/release/configuration/ldap/ldap_config.py#L99).

Comment (Sispheor) [reply Squest/ldap_config.py:21]: Same comment for all other fields actually. Like TLS config.

File: docker/scripts/mariadb-init.sh (hunk 6 above)
Comment (Sispheor) [reply docker/scripts/mariadb-init.sh:1]: All good then.

File: docker/environment_variables/psql.env [hunk 12]
```diff
@@ -0,0 +1,8 @@
+DB_HOST=db
+
+# PSQL data
+DB_DATABASE=squest
+DB_PASSWORD=squest
+DB_USER=squest
+POSTGRES_PASSWORD=psqlSuperP@ssword
```
Comment (Sispheor) [inline docker/environment_variables/psql.env:7]: Maybe here we should keep only one file named db.env and use generic variable name like you did.
The only missing ones are MYSQL_ROOT_PASSWORD and POSTGRES_PASSWORD which are the same and could be renamed "DB_ADMIN_PASSWORD" or "DB_ROOT_PASSWORD".
Then the mapping is done in the compose file.
 What do you think?

File: docker/entrypoint.sh [hunk 13]
```diff
@@ -1,10 +1,11 @@
 #!/usr/bin/env bash
+ENV DATABASE=""
```
Comment (Sispheor) [inline docker/entrypoint.sh:2]: Could be placed in the squest.env file, right?

File: docs/dev/dev-env.md (hunk 9 above)
Comment (clementTal) [reply docs/dev/dev-env.md]: I've fixed it for mysql, but leaved it as is for psql as it is not the default behavior.

File: Squest/ldap_config.py (hunk 11 above)
Comment (clementTal) [reply Squest/ldap_config.py:21]: leaved the AUTH_LDAP_CONNECTION_OPTIONS as is but wrapped all others feilds, and add a bit of documentation.

File: .dockerignore (hunk 1 above)
Comment (clementTal) [reply .dockerignore:4]: done

File: Squest/ldap_config.py (hunk 4 above)
Comment (clementTal) [reply Squest/ldap_config.py:6]: Done

File: docs/dev/dev-env.md [hunk 14]
```diff
@@ -38,7 +38,11 @@ The development environment is composed of 4 parts:
 
 Run the Docker compose file with only required services to bring up database, message broker and other required system
 ```bash
-docker-compose -f docker-compose.yml -f dev.docker-compose.yml up db phpmyadmin rabbitmq redis-cache
+docker-compose -f docker-compose.yml  -f dev.docker-compose.yml up db phpmyadmin rabbitmq redis-cache
+```
+If you use postgreSQL, update docker/environment_variables/squest.env, then run
+```bash
+docker-compose -f docker-compose.yml  -f psql.docker-composer.yaml -f dev.docker-compose.yml up db phpmyadmin rabbitmq redis-cache
```
Comment (Sispheor) [inline docs/dev/dev-env.md:45]: Shouldn't we use the name `psql.docker-compose.yml` instead of `psql.docker-composer.yaml` to be consistent?

**Review by Sispheor — CHANGES_REQUESTED** [review]
Just a file name to align and it's all good to me.
@EliasBoulharts @a-belhadj if you have something to add don't hesitate.

@clementTal thanks again for your contribution 🙂

**Comment by Sispheor** [comment]
Tests are failing. You need to ajust the poetry version in the CI [here](https://github.com/HewlettPackard/squest/blob/master/.github/workflows/on-pull-request-workflow.yml#L15)

File: poetry.lock [hunk 15]
```diff
@@ -1,24 +1,34 @@
+# This file is automatically @generated by Poetry and should not be changed by hand.
+
 [[package]]
 name = "amqp"
-version = "5.0.9"
```
Comment (Sispheor) [inline poetry.lock:3]: Wooo, I haven't noticed at first that this file changed a lot.
Why did you bump all lib?
Lib bump should not be part of this PR.
Could you please reset this file and add your required psql lib?
Thx

**Review by Sispheor — CHANGES_REQUESTED** [review]
Please reset the poetry config to the original one.

File: poetry.lock (hunk 15 above)
Comment (clementTal) [reply poetry.lock:3]: Just saw that. I'm not an poetry expert, maybe you should fix versions on pyproject.toml ? (only minimal version are set using "^").
Anyway, I'll revert poetry.lock today

**Comment by Sispheor** [comment]
It's  weird that you got this behavior, as it's the main purpose of poetry to align developers on same versions. 
It should not try to bump automatically libs.
Anyway. Thanks for the fix. I'll wait for the test to pass and merge.
It will be a part of the next release.

**Review by a-belhadj — APPROVED** [review]

**Comment by clementTal** [comment]
> It's weird that you got this behavior, as it's the main purpose of poetry to align developers on same versions. It should not try to bump automatically libs. Anyway. Thanks for the fix. I'll wait for the test to pass and merge. It will be a part of the next release.

It's juste because I ran `poetry lock` without the `--no-update`. As dependencies version in the `pyproject.toml` have the `^X.Y.Z` descriptor, poetry update this dependancies to the last minor version: `An update is allowed if the new version number does not modify the left-most non-zero digit in the major, minor, patch grouping`. 

Thank for reviewing adding my changes :)
